### DISCUSSION TITLE
Daily Total kWh consumed by Heatpump / Writing back solar to Heatpump

### DISCUSSION THREAD

**Discussion by frederikbeekman** [kodebach/hacs-idm-heatpump#144] [Q&A] [answered]
Hola! 

Thanks for building this, really appreciated!!! :) Question: Could you include a daily total kWh consumed by the Heatpump for in the Energy Dashboard?

Also as another request:
Could you create functionality to write back the solar total (register78) and excess solar (register74) back to the heat pump? I use a separate script for that now, would be great if I can get that done through this automation too!

I have 2 IDM iPumps with Navigator 2.0 and it works like a charm!

**Comment by kodebach** [comment]
> Could you create functionality to write back the solar total (register78) and excess solar (register74) back to the heat pump?

That is already supported via the `set_power` service (now also called "action" by Home Assistant) provided by the integration.

> Could you include a daily total kWh consumed by the Heatpump for in the Energy Dashboard?

AFAIK the Energy Dashboard actually wants a monotonically increasing total energy sensor (one that never resets) and it calculates daily changes from that. You can achieve that by using the [integral helper](https://www.home-assistant.io/integrations/integration/) with the "Aktuelle Leistungsaufnahme" sensor as input.

If you also want a daily total sensor, you should be able to do this via the [utility meter integration](https://www.home-assistant.io/integrations/utility_meter/).

I'm not keen on implementing such a features directly because then people will also request monthly/yearly totals and other configurations and it is already possibly by combining integrations.

**Comment by frederikbeekman** [comment]
Hi Thanks for your quick reply! Noted about the answers, very nice! I will setup some sensors myself, that should be doable ;)

With regards to the set_power function: This actually results in an error for me:
![Screenshot 2024-09-30 at 08 12 00](https://github.com/user-attachments/assets/1afe6768-b306-4324-9d24-0014b8c89837)

And then the Error Message (KeyError)
![Screenshot 2024-09-30 at 08 13 16](https://github.com/user-attachments/assets/61cfa4ed-b988-4f47-8412-7c90c789037d)

Am I doing something wrong here? Thanks again for your help!

**Reply by kodebach** [reply, marked as answer]
No, when using the UI Mode to invoke the action you can't really do anything wrong (*). Seems there is a bug, I'll take look, follow #147 for updates.


(*) Other than selecting the wrong target, but that would show up as "Service not supported".

**Comment by 5erv3** [comment]
Hi @kodebach,
thanks for the great tool first of all!
Using the integral helper with "Aktuelle Leistungsaufnahme" does not really work well unfortunately:
Because the max refresh interval of 1min there is a huge error in the total sum. The calculated overall consumption is way higher than displayed in the IDM portal. 
Imagine the pump runs @6kW = 6000W/60min = 100W/min. If the power consumption of the pump changes faster than 1min (which it does very often in my case) the error gets huge. I tried it for some time in the energy dashboard and got an error of >30%.

Do you think it is possible to get the total energy consumption somehow? Then the update interval would not be that important.

**Reply by frederikbeekman** [reply]
This is what im seeing now! I created a utility meter, but as you can see below, the numbers are way, way too high as you indicated... It would be great if we could get the total energy consumption per heat pump!
![Screenshot 2024-10-01 at 11 17 25](https://github.com/user-attachments/assets/61143593-b765-450c-bcff-1c1d903b15d7)

![Screenshot 2024-10-01 at 11 17 46](https://github.com/user-attachments/assets/31c4e339-3c87-4095-8864-d391e6d4edc4)

**Reply by kodebach** [reply]
Ah true, the doing integral in HA is of course imprecise. Interestingly, I do not have this "Electrical Power Consumption" box in my IDM interface either. It could be that a different box ("Stromfluss") replaces it, because I have connected a Fronius solar inverter. However, this "Stromfluss" UI seems to not actually show the total consumption over the lifetime either, and only shows the consumption since the inverter was connected. At least it seems that way, since all the production and consumption values add up to zero.

The problem is, AFAICT there is no way to read these values (from either type of UI) via Modbus. "Aktuelle Leistungsaufnahme" is the only register for an electrical value I found in the documentation. All the other `kW` and `kWh` values are clearly heat-energy-based values. In the "GLT Monitor" in the IDM Web UI I at somepoint also found register 4124 "elektrische Gesamtleistung". But since the Web UI doesn't show the units, I haven't integrated this yet.

If anybody has a more up-to-date version of IDM's Modbus TCP documentation, which mentions these values, it would help greatly.

**Reply by 5erv3** [reply]
I also did not find anything in the TCP documentation I found. It might help to decrease the interval to 1s. @kodebach do you think this is a problem? 
Maybe the code can be adjusted to just update the "Aktuelle Leistungsaufnahme" in 1s interval and leave the rest on the configured value? I might have a look when I find time.

**Reply by kodebach** [reply]
I added the limit because of this issue https://github.com/kodebach/hacs-idm-heatpump/issues/16. I'm not entirely sure, how _reading_ too fast/often could damage the hardware, but I wanted to be safe. I know that IDM's Windows Software can display sensor data in almost real time, so there definitely is some safe way to get the data faster. However, the Windows Software doesn't use Modbus TCP (it works without enabling Modbus), so you'd have to roll your own setup for using that.
